singularity-eos/eos/eos_base.hpp

```diff
@@ -0,0 +1,330 @@
+//------------------------------------------------------------------------------
+// © 2021. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_EOS_BASE_
+#define _SINGULARITY_EOS_EOS_EOS_BASE_
+
+#include <string>
+
+#include <ports-of-call/portability.hpp>
+
+// Macro for portableFor naming
+#define SG_MEMBER_FUNC_NAME() \
+  (std::string(typeid(CRTP).name()) \
+    + std::string("::") + std::string(__func__))
+
+namespace singularity {
+namespace eos_base {
+
+// This Macro adds the `using` statements that allow for the base class
+// vector functionality to overload the scalar implementations in the derived
+// classes
+#define SG_ADD_BASE_CLASS_USINGS(EOSDERIVED) \
+using EosBase<EOSDERIVED>::TemperatureFromDensityInternalEnergy;\
+using EosBase<EOSDERIVED>::InternalEnergyFromDensityTemperature;\
+using EosBase<EOSDERIVED>::PressureFromDensityTemperature;\
+using EosBase<EOSDERIVED>::PressureFromDensityInternalEnergy;\
+using EosBase<EOSDERIVED>::SpecificHeatFromDensityTemperature;\
+using EosBase<EOSDERIVED>::SpecificHeatFromDensityInternalEnergy;\
+using EosBase<EOSDERIVED>::BulkModulusFromDensityTemperature;\
+using EosBase<EOSDERIVED>::BulkModulusFromDensityInternalEnergy;\
+using EosBase<EOSDERIVED>::GruneisenParamFromDensityTemperature;\
+using EosBase<EOSDERIVED>::GruneisenParamFromDensityInternalEnergy;\
+using EosBase<EOSDERIVED>::PTofRE;\
+using EosBase<EOSDERIVED>::FillEos;
+
+/*
+This is a CRTP that allows for static inheritance so that default behavior for
```

Comment:
Probably worth defining CRTP in this comment and pointing to a wikipedia article or something.

Reply:
Added the Wiki article